src/bmat.h

```diff
@@ -0,0 +1,222 @@
+
+    //! Returns \c true if \c this is greater than \p that.
+    //!
+    //! This method checks whether a BMat8 objects is greater than another.
+    //! We order by the results of to_int() for each matrix.
+    bool operator>(BMat8 const &that) const { return _data > that._data; }
+
+    //! Returns the entry in the (\p i, \p j)th position.
+    //!
+    //! This method returns the entry in the (\p i, \p j)th position.
+    //! Note that since all matrices are internally represented as 8x8, it
+    //! is possible to access entries that you might not believe exist.
+    bool operator()(size_t i, size_t j) const {
+      LIBSEMIGROUPS_ASSERT(0 <= i && i < 8);
+      LIBSEMIGROUPS_ASSERT(0 <= j && j < 8);
+      return (_data << (8 * i + j)) >> 63;
+    }
+
+    //! Returns the integer representation of \c this.
+    //!
+    //! Returns an unsigned integer obtained by interpreting an 8x8
+    //! BMat8 as a sequence of 64 bits (reading rows left to right,
+    //! from top to bottom)and then this sequence as an unsigned int.
+    inline uint64_t to_int() const { return _data; }
+
+    //! Returns the transpose of \c this
+    //!
+    //! Returns the standard matrix transpose of a BMat8.
+    inline BMat8 transpose() const {
+      uint64_t x = _data;
+      uint64_t y = (x ^ (x >> 7)) & 0xAA00AA00AA00AA;
+      x = x ^ y ^ (y << 7);
+      y = (x ^ (x >> 14)) & 0xCCCC0000CCCC;
+      x = x ^ y ^ (y << 14);
+      y = (x ^ (x >> 28)) & 0xF0F0F0F0;
+      x = x ^ y ^ (y << 28);
+      return BMat8(x);
+    }
+
+    //! Returns the matrix product of \c this and \p that
+    //!
+    //! This method returns the standard matrix product (over the
+    //! boolean semiring) of two BMat8.
+    // https://stackoverflow.com/a/18448513
+    inline BMat8 operator*(BMat8 const &that) const {
+      uint64_t y = that.transpose()._data;
+      uint64_t data = 0;
+      uint64_t tmp = 0;
+      uint64_t diag = 0x8040201008040201;
+      for (int i = 0; i < 8; ++i) {
+        tmp = _data & y;
+        tmp |= tmp >> 1;
+        tmp |= tmp >> 2;
+        tmp |= tmp >> 4;
+        tmp &= 0x0101010101010101;
+        tmp *= 255;
+        tmp &= diag;
+        data |= tmp;
+        y = cyclic_shift(y);
+        tmp = 0;
+        diag = cyclic_shift(diag);
+      }
+      return BMat8(data);
+    }
+
+    //! Returns the identity BMat8
+    //!
+    //! This method returns the 8x8 BMat8 with 1s on the main diagonal.
+    inline BMat8 one() const { return BMat8(0x8040201008040201); }
+
+    //! Insertion operator
+    //!
+    //! This method allows BMat8s to be inserted into a stream.
+    friend std::ostream &operator<<(std::ostream &os, BMat8 const &bm) {
+      os << bm.to_string();
+      return os;
+    }
+
+    //! Returns a string representation of \c this
+    //!
+    //! This method returns a string which represents a BMat8, which may
+    //! for example be used to display the BMat8.
+    std::string to_string() const;
```

Comment:
I think we should replace this with a method for the `operator<<` of a `std::ostringstream`, and a `BMat8`.